extmod/moducryptolib.c

```diff
@@ -356,9 +361,190 @@ STATIC const mp_obj_type_t ucryptolib_aes_type = {
     .locals_dict = (void *)&ucryptolib_aes_locals_dict,
 };
 
+#if MICROPY_PY_UCRYPTOLIB_GCM
+
+#if MICROPY_SSL_AXTLS
+#warning GCM is not currentl implementing with axTLS
+#endif
+
+#if MICROPY_SSL_MBEDTLS
+typedef struct _mp_obj_aesgcm_t {
+    mp_obj_base_t base;
+    int key_len;
+    uint8_t key[32];
+} mp_obj_aesgcm_t;
+
+STATIC mp_obj_t ucryptolib_aesgcm_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+
+    mp_obj_aesgcm_t *o = m_new_obj(mp_obj_aesgcm_t);
+
+    o->base.type = type;
+    mp_buffer_info_t keyinfo;
+    mp_get_buffer_raise(args[0], &keyinfo, MP_BUFFER_READ);
+    if (32 != keyinfo.len && 16 != keyinfo.len) {
+        mp_raise_ValueError(MP_ERROR_TEXT("key length"));
+    }
+
+    o->key_len = keyinfo.len;
+    memcpy(&o->key, keyinfo.buf, keyinfo.len);
+
+    return MP_OBJ_FROM_PTR(o);
+}
+
+STATIC mp_obj_t ucryptolib_aesgcm_enc_dec_check_args(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs,
+    mp_buffer_info_t *nonce, mp_buffer_info_t *input, mp_buffer_info_t *output, mp_buffer_info_t *ad, int in_out_delta) {
+    mp_obj_t out_obj;
+    mp_map_elem_t *map_element;
+
+    if (n_args != 3) {
+        mp_raise_ValueError(MP_ERROR_TEXT("2 positional arguments required"));
+    }
+
+    mp_get_buffer_raise(args[1], nonce, MP_BUFFER_READ);
+    mp_get_buffer_raise(args[2], input, MP_BUFFER_READ);
+
+    size_t output_len = input->len + in_out_delta;
+
+    map_element = mp_map_lookup(kwargs, MP_OBJ_NEW_QSTR(MP_QSTR_out_buf), MP_MAP_LOOKUP);
+    if (map_element != NULL) {
+        out_obj = map_element->value;
+        mp_get_buffer_raise(out_obj, output, MP_BUFFER_WRITE);
+        if (output->len != output_len) {
+            mp_raise_ValueError(MP_ERROR_TEXT("incorrect output length"));
+        }
+    } else {
+        out_obj = MP_OBJ_NULL;
+        output->buf = m_new(char, output_len);
+        output->len = output_len;
+    }
+
+    map_element = mp_map_lookup(kwargs, MP_OBJ_NEW_QSTR(MP_QSTR_adata), MP_MAP_LOOKUP);
+    if (map_element != NULL) {
+        mp_get_buffer_raise(map_element->value, ad, MP_BUFFER_READ);
+    } else {
+        ad->buf = NULL;
+        ad->len = 0;
+    }
+
+    // Maybe this should check for spurious keyword args...
```

Comment:
The associated data needs to be optional too, so that would need to be:
`encrypt(nonce, data, [associated_data, [out_buf]])` but yes, that would remove the need for keywords.

Reply:
I think we can stick with [cryptography's interface](https://github.com/pyca/cryptography/blob/3367c18bf2e71639843e38498f5ad2159835122d/src/cryptography/hazmat/primitives/ciphers/aead.py#L89), where it's not optional, but allowed to be `None` or `b""`.

Reply:
That works too, although by the time I have (a) moved to positional arguments, (b) moved the AD to be third and (c) added an extra argument that's not in `crytpography`'s, it seems like I might as well make the AD optional to save Micropython users some typing!

Reply:
Well, that works, though I prefer sticking with the most "standard" interface possible, in this case cryptography's.

I guess that this logic
```
if len(args) >= 3:
    ad = args[2]
    if len(args) == 4:
        out = args[3]
```
is god as well.